File: hera/services/market.ts
/**
 * Persists the market engine's state and drives ticks for a guild.
 *
 * One engine instance is kept per guild so multiple servers can run independent
 * markets without interfering with each other.
 */
import type { MarketConfig } from '../config';
import type { Database } from '../database';
import { COMPANIES } from '../market/companies';
import { CompanyState, MarketEngine, type NewsItem, type TickResult } from '../market/engine';
import type { Random } from '../market/random';

export interface MarketSnapshot {
  tick: number;
  regime: string;
  indexValue: number;
  indexChange: number;
  companies: CompanyState[];
}

interface CompanyRow {
  symbol: string;
  name: string;
  sector: string;
  description: string;
  price: number;
  previous_close: number;
  open_price: number;
  day_high: number;
  day_low: number;
  volatility: number;
  drift: number;
  beta: number;
  shares_outstanding: number;
  dividend_yield: number;
  halted_until_tick: number;
  active_event: string | null;
  event_ticks_left: number;
  event_magnitude: number;
  last_dividend_tick: number;
}

interface MarketStateRow {
  tick: number;
  regime: string;
  regime_ticks_left: number;
  index_value: number;
  previous_index: number;
}

export class MarketService {
  private readonly engines = new Map<string, MarketEngine>();

  constructor(
    private readonly db: Database,
    public readonly config: MarketConfig,
    private readonly rng?: Random,
  ) {}

  // ------------------------------------------------------------------ setup

  getEngine(guildId: string): MarketEngine {
    const existing = this.engines.get(guildId);
    if (existing) return existing;
    const engine = new MarketEngine(this.config, { rng: this.rng });
    this.load(guildId, engine);
    this.engines.set(guildId, engine);
    return engine;
  }

  private load(guildId: string, engine: MarketEngine): void {
    let rows = this.db.fetchall<CompanyRow>('SELECT * FROM companies WHERE listed = 1 ORDER BY symbol');
    if (rows.length === 0) {
      this.seedCompanies();
      rows = this.db.fetchall<CompanyRow>('SELECT * FROM companies WHERE listed = 1 ORDER BY symbol');
    }

    const companies: Record<string, CompanyState> = {};
    for (const row of rows) {
      const base = engine.companies[row.symbol];
      if (!base) continue;
      const state = new CompanyState({
        symbol: row.symbol,
        name: row.name,
        sector: row.sector,
        description: row.description,
        price: Number(row.price),
        volatility: Number(row.volatility),
        drift: Number(row.drift),
        beta: Number(row.beta),
        sharesOutstanding: Number(row.shares_outstanding),
        dividendYield: Number(row.dividend_yield),
      });
      state.previousClose = Number(row.previous_close);
      state.openPrice = Number(row.open_price);
      state.dayHigh = Number(row.day_high);
      state.dayLow = Number(row.day_low);
      state.fairValue = Number(row.price);
      state.haltedUntilTick = Number(row.halted_until_tick);
      state.activeEvent = row.active_event;
      state.eventTicksLeft = Number(row.event_ticks_left);
      state.eventMagnitude = Number(row.event_magnitude);
      state.lastDividendTick = Number(row.last_dividend_tick);
      companies[row.symbol] = state;
    }

    const stateRow = this.db.fetchone<MarketStateRow>(
      'SELECT * FROM market_state WHERE guild_id = ?',
      [guildId],
    );
    const tick = stateRow ? Number(stateRow.tick) : 0;
    const regime = stateRow ? stateRow.regime : 'neutral';
    const regimeTicksLeft = stateRow ? Number(stateRow.regime_ticks_left) : 40;
    const indexValue = stateRow ? Number(stateRow.index_value) : this.config.startingIndex;
    const previousIndex = stateRow ? Number(stateRow.previous_index) : indexValue;

    engine.loadState({
      tick,
      regime,
      regimeTicksLeft,
      indexValue,
      previousIndex,
      companies,
    });
    engine.previousIndex = previousIndex;
  }

  private seedCompanies(): void {
    this.db.transaction(() => {
      this.db.executemany(
        `INSERT INTO companies (
            symbol, name, sector, description, price, previous_close,
            open_price, day_high, day_low, volatility, drift, beta,
            shares_outstanding, market_cap, dividend_yield
         ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
         ON CONFLICT (symbol) DO NOTHING`,
        COMPANIES.map((seed) => [
          seed.symbol,
          seed.name,
          seed.sector,
          seed.description,
          seed.price,
          seed.price,
          seed.price,
          seed.price,
          seed.price,
          seed.volatility,
          seed.drift,
          seed.beta,
          seed.sharesOutstanding,
          seed.price * seed.sharesOutstanding,
          seed.dividendYield ?? 0.0,
        ]),
      );
    });
  }

  // ------------------------------------------------------------------- ticks

  tick(guildId: string, options: { newDay?: boolean } = {}): TickResult {
    const engine = this.getEngine(guildId);
    const result = engine.advance({ newDay: options.newDay ?? false });
    this.persist(guildId, engine, result);
    return result;
  }

  private persist(guildId: string, engine: MarketEngine, result: TickResult): void {
    this.db.transaction(() => {
      this.db.executemany(
        `UPDATE companies SET
            price = ?, previous_close = ?, open_price = ?, day_high = ?, day_low = ?,
            market_cap = ?, halted_until_tick = ?, active_event = ?, event_ticks_left = ?,
            event_magnitude = ?, last_dividend_tick = ?
         WHERE symbol = ?`,
        Object.values(engine.companies).map((company) => [
          company.price,
          company.previousClose,
          company.openPrice,
          company.dayHigh,
          company.dayLow,
          company.marketCap,
          company.haltedUntilTick,
          company.activeEvent,
          company.eventTicksLeft,
          company.eventMagnitude,
          company.lastDividendTick,
          company.symbol,
        ]),
      );
      this.db.executemany(
        `INSERT INTO price_history (symbol, tick, price) VALUES (?, ?, ?)
         ON CONFLICT (symbol, tick) DO UPDATE SET price = excluded.price`,
        Object.values(engine.companies).map((company) => [company.symbol, engine.tick, company.price]),
      );
      this.db.execute(
        `INSERT INTO market_state
            (guild_id, tick, regime, regime_ticks_left, index_value, previous_index, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, unixepoch('subsec'))
         ON CONFLICT (guild_id) DO UPDATE SET
            tick = excluded.tick,
            regime = excluded.regime,
            regime_ticks_left = excluded.regime_ticks_left,
            index_value = excluded.index_value,
            previous_index = excluded.previous_index,
            updated_at = excluded.updated_at`,
        [
          guildId,
          engine.tick,
          engine.regime,
          engine.regimeTicksLeft,
          engine.indexValue,
          engine.previousIndex,
        ],
      );
      if (result.news.length > 0) {
        this.db.executemany(
          `INSERT INTO market_news (guild_id, tick, symbol, headline, impact) VALUES (?, ?, ?, ?, ?)`,
          result.news.map((item) => [guildId, engine.tick, item.symbol, item.headline, item.impact]),
        );
      }
    });
    this.pruneHistory(engine.tick);
  }

  /** Keep the history table bounded so the file does not grow forever. */
  private pruneHistory(tick: number): void {
    const cutoff = tick - this.config.historyLimit;
    if (cutoff <= 0) return;
    this.db.execute('DELETE FROM price_history WHERE tick < ?', [cutoff]);
  }

  // ------------------------------------------------------------------ queries

  snapshot(guildId: string): MarketSnapshot {
    const engine = this.getEngine(guildId);
    const indexChange =
      engine.previousIndex === 0
        ? 0.0
        : (engine.indexValue - engine.previousIndex) / engine.previousIndex;
    return {
      tick: engine.tick,
      regime: engine.regime,
      indexValue: engine.indexValue,
      indexChange,
      companies: engine.snapshot(),
    };
  }

  getCompany(guildId: string, symbol: string): CompanyState | undefined {
    return this.getEngine(guildId).companies[symbol.toUpperCase()];
  }

  /** Accept a ticker or a company name (prefix match) and return the match. */
  resolveSymbol(guildId: string, query: string): CompanyState | undefined {
    const engine = this.getEngine(guildId);
    const normalized = query.trim().toUpperCase();
    if (!normalized) return undefined;
    if (engine.companies[normalized]) return engine.companies[normalized];
    const companies = Object.values(engine.companies);
    for (const company of companies) {
      if (company.name.toUpperCase().startsWith(normalized)) return company;
    }
    for (const company of companies) {
      if (company.name.toUpperCase().includes(normalized)) return company;
    }
    return undefined;
  }

  history(symbol: string, limit = 100): [number, number][] {
    const rows = this.db.fetchall<{ tick: number; price: number }>(
      'SELECT tick, price FROM price_history WHERE symbol = ? ORDER BY tick DESC LIMIT ?',
      [symbol.toUpperCase(), limit],
    );
    return rows.reverse().map((row) => [Number(row.tick), Number(row.price)]);
  }

  news(guildId: string, limit = 8): NewsItem[] {
    const rows = this.db.fetchall<{ symbol: string | null; headline: string; impact: number }>(
      'SELECT symbol, headline, impact FROM market_news WHERE guild_id = ? ORDER BY id DESC LIMIT ?',
      [guildId, limit],
    );
    return rows.map((row) => ({
      symbol: row.symbol,
      headline: row.headline,
      impact: Number(row.impact),
    }));
  }

  /** Reconstruct the index from per-symbol history at each recorded tick. */
  indexHistory(guildId: string, limit = 100): [number, number][] {
    const rows = this.db.fetchall<{ tick: number; cap: number }>(
      `SELECT tick, SUM(h.price * c.shares_outstanding) AS cap
       FROM price_history h JOIN companies c ON c.symbol = h.symbol
       GROUP BY tick ORDER BY tick DESC LIMIT ?`,
      [limit],
    );
    if (rows.length === 0) return [];
    const caps = rows.reverse().map((row) => [Number(row.tick), Number(row.cap)] as [number, number]);
    const baseline = caps[0][1] || 1.0;
    return caps.map(([tick, cap]) => [tick, (this.config.startingIndex * cap) / baseline]);
  }
}

File: hera/services/market.py
"""Persists the market engine's state and drives ticks for a guild."""

from __future__ import annotations

import random
from dataclasses import dataclass

from ..config import MarketConfig
from ..database import Database
from ..market.companies import COMPANIES
from ..market.engine import CompanyState, MarketEngine, NewsItem, TickResult


@dataclass
class MarketSnapshot:
    tick: int
    regime: str
    index_value: float
    index_change: float
    companies: list[CompanyState]


class MarketService:
    """Owns ``companies``, ``price_history``, ``market_state`` and ``market_news``.

    One engine instance is kept per guild so multiple servers can run
    independent markets without interfering with each other.
    """

    def __init__(
        self,
        db: Database,
        market_config: MarketConfig,
        *,
        rng: random.Random | None = None,
    ) -> None:
        self.db = db
        self.config = market_config
        self._engines: dict[int, MarketEngine] = {}
        self._rng = rng

    # ------------------------------------------------------------------ setup

    async def get_engine(self, guild_id: int) -> MarketEngine:
        if guild_id in self._engines:
            return self._engines[guild_id]
        engine = MarketEngine(self.config, rng=self._rng)
        await self._load(guild_id, engine)
        self._engines[guild_id] = engine
        return engine

    async def _load(self, guild_id: int, engine: MarketEngine) -> None:
        rows = await self.db.fetchall(
            "SELECT * FROM companies WHERE listed = 1 ORDER BY symbol"
        )
        if not rows:
            await self._seed_companies()
            rows = await self.db.fetchall(
                "SELECT * FROM companies WHERE listed = 1 ORDER BY symbol"
            )

        companies: dict[str, CompanyState] = {}
        for row in rows:
            symbol = row["symbol"]
            base = engine.companies.get(symbol)
            if base is None:
                continue
            state = CompanyState(
                symbol=symbol,
                name=row["name"],
                sector=row["sector"],
                description=row["description"],
                price=float(row["price"]),
                previous_close=float(row["previous_close"]),
                open_price=float(row["open_price"]),
                day_high=float(row["day_high"]),
                day_low=float(row["day_low"]),
                fair_value=float(row["price"]),
                volatility=float(row["volatility"]),
                drift=float(row["drift"]),
                beta=float(row["beta"]),
                shares_outstanding=int(row["shares_outstanding"]),
                dividend_yield=float(row["dividend_yield"]),
                halted_until_tick=int(row["halted_until_tick"]),
                active_event=row["active_event"],
                event_ticks_left=int(row["event_ticks_left"]),
                event_magnitude=float(row["event_magnitude"]),
                last_dividend_tick=int(row["last_dividend_tick"]),
            )
            companies[symbol] = state

        state_row = await self.db.fetchone(
            "SELECT * FROM market_state WHERE guild_id = ?", (guild_id,)
        )
        tick = int(state_row["tick"]) if state_row else 0
        regime = state_row["regime"] if state_row else "neutral"
        regime_ticks_left = int(state_row["regime_ticks_left"]) if state_row else 40
        index_value = float(state_row["index_value"]) if state_row else self.config.starting_index
        previous_index = float(state_row["previous_index"]) if state_row else index_value

        engine.load_state(
            tick=tick,
            regime=regime,
            regime_ticks_left=regime_ticks_left,
            index_value=index_value,
            previous_index=previous_index,
            companies=companies,
        )
        engine.previous_index = previous_index
        # Re-anchor the index baseline so a restart does not produce a phantom jump.
        engine._initial_market_cap = sum(c.market_cap for c in companies.values()) or 1.0

    async def _seed_companies(self) -> None:
        async with self.db.transaction() as conn:
            await conn.executemany(
                """
                INSERT INTO companies (
                    symbol, name, sector, description, price, previous_close,
                    open_price, day_high, day_low, volatility, drift, beta,
                    shares_outstanding, market_cap, dividend_yield
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT (symbol) DO NOTHING
                """,
                [
                    (
                        seed.symbol,
                        seed.name,
                        seed.sector,
                        seed.description,
                        seed.price,
                        seed.price,
                        seed.price,
                        seed.price,
                        seed.price,
                        seed.volatility,
                        seed.drift,
                        seed.beta,
                        seed.shares_outstanding,
                        seed.price * seed.shares_outstanding,
                        seed.dividend_yield,
                    )
                    for seed in COMPANIES
                ],
            )

    # ------------------------------------------------------------------- ticks

    async def tick(self, guild_id: int, *, new_day: bool = False) -> TickResult:
        engine = await self.get_engine(guild_id)
        result = engine.advance(new_day=new_day)
        await self._persist(guild_id, engine, result)
        return result

    async def _persist(self, guild_id: int, engine: MarketEngine, result: TickResult) -> None:
        async with self.db.transaction() as conn:
            await conn.executemany(
                """
                UPDATE companies SET
                    price = ?, previous_close = ?, open_price = ?, day_high = ?,
                    day_low = ?, market_cap = ?, halted_until_tick = ?,
                    active_event = ?, event_ticks_left = ?, event_magnitude = ?,
                    last_dividend_tick = ?
                WHERE symbol = ?
                """,
                [
                    (
                        company.price,
                        company.previous_close,
                        company.open_price,
                        company.day_high,
                        company.day_low,
                        company.market_cap,
                        company.halted_until_tick,
                        company.active_event,
                        company.event_ticks_left,
                        company.event_magnitude,
                        company.last_dividend_tick,
                        company.symbol,
                    )
                    for company in engine.companies.values()
                ],
            )
            await conn.executemany(
                """
                INSERT INTO price_history (symbol, tick, price) VALUES (?, ?, ?)
                ON CONFLICT (symbol, tick) DO UPDATE SET price = excluded.price
                """,
                [(c.symbol, engine.tick, c.price) for c in engine.companies.values()],
            )
            await conn.execute(
                """
                INSERT INTO market_state
                    (guild_id, tick, regime, regime_ticks_left, index_value, previous_index,
                     updated_at)
                VALUES (?, ?, ?, ?, ?, ?, unixepoch('subsec'))
                ON CONFLICT (guild_id) DO UPDATE SET
                    tick = excluded.tick,
                    regime = excluded.regime,
                    regime_ticks_left = excluded.regime_ticks_left,
                    index_value = excluded.index_value,
                    previous_index = excluded.previous_index,
                    updated_at = excluded.updated_at
                """,
                (
                    guild_id,
                    engine.tick,
                    engine.regime,
                    engine.regime_ticks_left,
                    engine.index_value,
                    engine.previous_index,
                ),
            )
            if result.news:
                await conn.executemany(
                    """
                    INSERT INTO market_news (guild_id, tick, symbol, headline, impact)
                    VALUES (?, ?, ?, ?, ?)
                    """,
                    [
                        (guild_id, engine.tick, item.symbol, item.headline, item.impact)
                        for item in result.news
                    ],
                )
        await self._prune_history(engine.tick)

    async def _prune_history(self, tick: int) -> None:
        """Keep the history table bounded so the file does not grow forever."""
        cutoff = tick - self.config.history_limit
        if cutoff <= 0:
            return
        await self.db.execute("DELETE FROM price_history WHERE tick < ?", (cutoff,))

    # ------------------------------------------------------------------ queries

    async def snapshot(self, guild_id: int) -> MarketSnapshot:
        engine = await self.get_engine(guild_id)
        index_change = (
            0.0
            if engine.previous_index == 0
            else (engine.index_value - engine.previous_index) / engine.previous_index
        )
        return MarketSnapshot(
            tick=engine.tick,
            regime=engine.regime,
            index_value=engine.index_value,
            index_change=index_change,
            companies=engine.snapshot(),
        )

    async def get_company(self, guild_id: int, symbol: str) -> CompanyState | None:
        engine = await self.get_engine(guild_id)
        return engine.companies.get(symbol.upper())

    async def resolve_symbol(self, guild_id: int, query: str) -> CompanyState | None:
        """Accept a ticker or a company name (prefix match) and return the match."""
        engine = await self.get_engine(guild_id)
        query = query.strip().upper()
        if not query:
            return None
        if query in engine.companies:
            return engine.companies[query]
        for company in engine.companies.values():
            if company.name.upper().startswith(query):
                return company
        for company in engine.companies.values():
            if query in company.name.upper():
                return company
        return None

    async def history(self, symbol: str, limit: int = 100) -> list[tuple[int, float]]:
        rows = await self.db.fetchall(
            "SELECT tick, price FROM price_history WHERE symbol = ? ORDER BY tick DESC LIMIT ?",
            (symbol.upper(), limit),
        )
        return [(int(row["tick"]), float(row["price"])) for row in reversed(rows)]

    async def news(self, guild_id: int, limit: int = 8) -> list[NewsItem]:
        rows = await self.db.fetchall(
            """
            SELECT symbol, headline, impact FROM market_news
            WHERE guild_id = ? ORDER BY id DESC LIMIT ?
            """,
            (guild_id, limit),
        )
        return [
            NewsItem(symbol=row["symbol"], headline=row["headline"], impact=float(row["impact"]))
            for row in rows
        ]

    async def index_history(self, guild_id: int, limit: int = 100) -> list[tuple[int, float]]:
        """Reconstruct the index from per-symbol history at each recorded tick."""
        rows = await self.db.fetchall(
            """
            SELECT tick, SUM(h.price * c.shares_outstanding) AS cap
            FROM price_history h
            JOIN companies c ON c.symbol = h.symbol
            GROUP BY tick
            ORDER BY tick DESC
            LIMIT ?
            """,
            (limit,),
        )
        if not rows:
            return []
        caps = [(int(row["tick"]), float(row["cap"])) for row in reversed(rows)]
        baseline = caps[0][1] or 1.0
        return [(tick, self.config.starting_index * cap / baseline) for tick, cap in caps]
